lib/Support/Progress.cpp

```diff
@@ -0,0 +1,265 @@
+      Advancements[LastIndex] = 1.0;
+    } else if (Stack.Tasks[LastIndex]->totalSteps()) {
+      float Index = Stack.Tasks[LastIndex]->stepIndex();
+      Advancements[LastIndex] = Index / *Stack.Tasks[LastIndex]->totalSteps();
+    }
+
+    for (signed I = LastIndex - 1; I >= 0; --I) {
+      auto *Task = Stack.Tasks[I];
+      revng_assert(not Task->completed());
+
+      auto MaybeStepsCount = Task->totalSteps();
+      if (MaybeStepsCount) {
+        auto StepsCount = *MaybeStepsCount;
+        revng_assert(Task->stepIndex() < StepsCount);
+
+        float Addendum = 0.0;
+
+        // auto *InnerTask = Stack.Tasks[I + 1];
+        // if (InnerTask->isSingleSubtask())
+        if (Task->currentStepHasSingleSubtask())
+          Addendum = Advancements[I + 1];
+
+        Advancements[I] = (float(Task->stepIndex()) + Addendum) / StepsCount;
+      }
+
+      revng_assert(Advancements[I] <= 1.0);
+    }
+
+    std::string Buffer;
+    llvm::raw_string_ostream Line(Buffer);
+
+    // Go back MaxProgressBars lines
+    Line << "\r\033[" << (MaxProgressBars) << "A";
+
+    for (size_t II = 0; II < MaxProgressBars; ++II) {
+      size_t I = MaxProgressBars - II - 1;
+      Line << "\r\033[2K";
+
+      if (I < Stack.Tasks.size()) {
+        auto *T = Stack.Tasks[I];
+        unsigned Percent = 100 * Advancements[I];
+        constexpr unsigned BarLength = 39;
+        unsigned Bar = BarLength * Advancements[I];
+
+        Line << "[";
+        unsigned I = 0;
+        for (; I < Bar; ++I) {
+          Line << "=";
+        }
+
+        if (I < BarLength) {
+          Line << ">";
+          ++I;
+        }
+
+        for (; I < BarLength; ++I) {
+          Line << " ";
+        }
+        Line << "] ";
+        Line << pad(llvm::Twine(Percent), 3) << "%";
+        Line << " " << T->name().str();
+
+        if (T->totalSteps())
+          Line << " (" << *T->totalSteps() << ")";
+
+        if (not T->stepName().empty())
+          Line << ": " << T->stepName().str();
+      }
+
+      Line << "\n";
+      Line.flush();
+    }
+
+    Output << Buffer;
+  }
+};
+
+using namespace llvm::cl;
+
+static auto RegisterTraceProgressListener = [](const std::string &Value) {
+  if (Value.size() > 0) {
+    llvm::ProgressReport->registerListener<TraceProgressListener>(Value);
+  }
+};
+
+static auto TPLCallback = callback(RegisterTraceProgressListener);
+
+static opt<std::string> TraceProgress("trace", TPLCallback);
+
+static auto RegisterTerminalBarsProgressListener = [](const bool &Value) {
+  using namespace llvm;
+  if (Value) {
+    ProgressReport->registerListener<TerminalBarsProgressListener>(errs());
+  }
+};
+
+static auto RTBPLCallback = callback(RegisterTerminalBarsProgressListener);
+
+static opt<bool> ProgressBars("progress", RTBPLCallback);
+
+static auto RegisterProgressPlain = callback([](const bool &Value) {
+  if (Value)
+    llvm::ProgressReport->registerListener<PlainProgressListener>(llvm::errs());
+});
```

Comment:
Can't you make these regular functions instead of lambdas so that you can avoid the awkward `callback` dance?

Reply:
I'll double check but IIRC the caller explicitly uses `operator()` instead of performing a simple function call, so regular functions won't work.